src/pineko/fonll.py

```diff
@@ -282,40 +179,77 @@ def produce_combined_fk(
     combined_fk.write_lz4(output_path_fk)
 
 
-def produce_ptos(fns, is_mixed_or_damp):
-    """Produce the list of PTOs needed for the requested fns."""
-    base_pto = FNS_BASE_PTO[fns]
-    # mixed FONLL
-    if fns in MIXED_ORDER_FNS:
-        return [
-            base_pto + 1,
-            base_pto + 1,
-            base_pto,
-            base_pto + 1,
-            base_pto + 1,
-            base_pto,
-            base_pto + 1,
-        ]
-    # plain FONLL with damping
-    elif is_mixed_or_damp:
-        return [base_pto for _ in range(7)]
-    # plain FONLL without damping
-    # In the case of not mixed and not damped FK tables, 5 sub FK tables are enough
-    else:
-        return [base_pto for _ in range(5)]
-
-
-def produce_fonll_tcards(tcard, tcard_parent_path, theoryid):
-    """Produce the seven fonll tcards from an original tcard.
-
-    The produced tcards are dumped in tcard_parent_path with names from '{theoryid}00.yaml' to '{theoryid}06.yaml'.
+FNS_BASE_PTO = {"FONLL-A": 1, "FONLL-B": 1, "FONLL-C": 2, "FONLL-D": 2, "FONLL-E": 3}
+MIXED_ORDER_FNS = ["FONLL-B", "FONLL-D"]
+
+
+@dataclasses.dataclass
+class SubTheoryConfig:
+    """Single (sub-)theory configuration."""
+
+    scheme: str
+    nf: int
+    parts: str
+    delta_pto: int = 0
+
+
+MIXED_FNS_CONFIG = [
+    SubTheoryConfig("FONLL-FFNS", 3, "full", 1),
+    SubTheoryConfig("FONLL-FFN0", 3, "full", 1),
+    SubTheoryConfig("FONLL-FFNS", 4, "massless"),
+    SubTheoryConfig("FONLL-FFNS", 4, "massive", 1),
+    SubTheoryConfig("FONLL-FFN0", 4, "full", 1),
+    SubTheoryConfig("FONLL-FFNS", 5, "massless"),
+    SubTheoryConfig("FONLL-FFNS", 5, "massive", 1),
+]
+"""Mixed FONLL schemes."""
+
+FNS_CONFIG = [
+    SubTheoryConfig("FONLL-FFNS", 3, "full"),
+    SubTheoryConfig("FONLL-FFN0", 3, "full"),
+    SubTheoryConfig("FONLL-FFNS", 4, "full"),
+    SubTheoryConfig("FONLL-FFN0", 4, "full"),
+    SubTheoryConfig("FONLL-FFNS", 5, "full"),
+]
+"""Plain FONLL schemes."""
```

Comment:
It's much tidier than before, and this I like a lot.

But it's still the full spell-out, and this I would avoid.

Leveraging the object, you could start removing some details. E.g. You just need a boolean flag to distinguish `FONLL-FFNS` from `FONLL-FFN0`, and make the string a `@property` (that essentially concatenates the fixed prefix `FONLL-FFN` with a single character, flag-dependent).

Even `full`, `massive`,  `massless`, are one out of 3, so it can be an enum. However, it's not only about making it an enum, but they can also be functions:
- `FNS` -> `full`
- `MIXED` -> `delta_pto = 1`
    - the second for each flavor is always `massive`, and the first `massless`, but for 3 flavors
    - this is the definition, and it'd scale for each number of flavors (that we do not need, but it would make the decision more transparent)

I'm not asking necessarily everything at once. But nFONLL is defined by some rules, and coding the rules in the algorithm is more transparent than the bare result.

Reply:
With all this, that the result of whatever function (possibly calling sub-functions) is going to be a `SubTheoryConfig` is a great idea. And (as said above) I like it a lot.

Reply:
> Leveraging the object, you could start removing some details. E.g. You just need a boolean flag to distinguish `FONLL-FFNS` from `FONLL-FFN0`, and make the string a `@property` (that essentially concatenates the fixed prefix `FONLL-FFN` with a single character, flag-dependent).

this was easy: b5674d0

the others are a bit more evolved ...

and particularly, I still believe there is something fishy with `is_mixed` and `is_damped` since they are strictly orthogonal ... and to be specific: it might well be that the name `MIXED_FNS_CONFIG` is simply misleading and it should be really `EXTENDED_CONFIG` which happens to be used for either damping or mixed is used (but has nothing to do with FNS)

Reply:
> it might well be that the name `MIXED_FNS_CONFIG` is simply misleading and it should be really `EXTENDED_CONFIG` which happens to be used for either damping or mixed is used (but has nothing to do with FNS)

Well, I acknowledge that it's not trivial, and I said it should not come necessarily now. But this kind of ambiguity would be almost over if the algorithm were transparent...

Reply:
let's keep further improvements for later then